tools/python/src/face_recognition.cpp

```diff
@@ -47,33 +48,78 @@ class face_recognition_model_v1
+            throw dlib::error("The array of images and the array of array of locations must be of the same size");
 
-        for (auto& f : faces)
+        int total_chips = 0;
+        for (auto& faces : batch_faces)
         {
-            if (f.num_parts() != 68 && f.num_parts() != 5)
-                throw dlib::error("The full_object_detection must use the iBUG 300W 68 point face landmark style or dlib's 5 point style.");
+            total_chips += faces.size();
+            for (auto& f : faces)
+            {
+                if (f.num_parts() != 68 && f.num_parts() != 5)
+                    throw dlib::error("The full_object_detection must use the iBUG 300W 68 point face landmark style or dlib's 5 point style.");
+            }
         }
 
 
-        std::vector<chip_details> dets;
-        for (auto& f : faces)
-            dets.push_back(get_face_chip_details(f, 150, 0.25));
         dlib::array<matrix<rgb_pixel>> face_chips;
-        extract_image_chips(img, dets, face_chips);
+        for (int i = 0; i < batch_imgs.size(); ++i)
+        {
+            auto& faces = batch_faces[i];
+            auto& img = batch_imgs[i];
+
+            std::vector<chip_details> dets;
+            for (auto& f : faces)
+                dets.push_back(get_face_chip_details(f, 150, 0.25));
+            dlib::array<matrix<rgb_pixel>> this_img_face_chips;
+            extract_image_chips(img, dets, this_img_face_chips);
 
-        std::vector<matrix<double,0,1>> face_descriptors;
-        face_descriptors.reserve(face_chips.size());
+            for (auto& chip : this_img_face_chips)
+                face_chips.push_back(chip);
+        }
 
+        std::vector<std::vector<matrix<double,0,1>>> face_descriptors(batch_imgs.size());
         if (num_jitters <= 1)
         {
             // extract descriptors and convert from float vectors to double vectors
-            for (auto& d : net(face_chips,16))
-                face_descriptors.push_back(matrix_cast<double>(d));
+            auto descriptors = net(face_chips, 16);
+            auto next = std::begin(descriptors);
+            for (int i = 0; i < batch_faces.size(); ++i)
+            {
+                for (int j = 0; j < batch_faces[i].size(); ++j)
+                {
+                    face_descriptors[i].push_back(matrix_cast<double>(*next++));
+                }
+            }
+            DLIB_ASSERT(next == std::end(descriptors));
         }
         else
         {
-            for (auto& fimg : face_chips)
-                face_descriptors.push_back(matrix_cast<double>(mean(mat(net(jitter_image(fimg,num_jitters),16)))));
+            // extract descriptors and convert from float vectors to double vectors
+            auto fimg = std::begin(face_chips);
+            for (int i = 0; i < batch_faces.size(); ++i)
+            {
+                for (int j = 0; j < batch_faces[i].size(); ++j)
+                {
+                    auto& r = matrix_cast<double>(mean(mat(net(
```

Comment:
You have `matrix_cast<double>` two times here.  It would also probably be faster if you did a std::move() on the insertion to the face_descriptors array.